[quant][graphmode][refactor] getQParams return a dictionary of qparams

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30553 [quant][graphmode] `insert_quant_dequant` pass support shared class types
* #30909 [quant][graphmode] Remove `fold_prepack` for now
* #30860 [quant][graphmode][refactor] Factor out getInvokedMethod in `InsertQuantDeQuantHelper`
* **#30859 [quant][graphmode][refactor] getQParams return a dictionary of qparams**

Summary:
We can return a dictionary of quantization parameters in getQParams
to simplify the code that uses the quantization parameters a bit.

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849023](https://our.internmc.facebook.com/intern/diff/D18849023)

jerryzh168 committed:
[quant][graphmode][refactor] getQParams return a dictionary of qparams

Summary:
We can dictionary of quantization parameters to simplify the code
handling these things a bit

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] getQParams return a dictionary of qparams"


Summary:
We can dictionary of quantization parameters to simplify the code
handling these things a bit

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jamesr66a commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -611,23 +611,21 @@ class InsertQuantDeQuantHelper {
-  std::tuple<IValue, IValue> getQParams(script::Module& module, Value* v);
+  // quantization parameters including scale, zero_point,
+  // scalar_type and axis(for per channel quantization)
+  std::unordered_map<std::string, IValue> getQParams(script::Module& module, Value* v);
```
To make things cleaner it might be good to add a `using` declaration for this `std::unordered_map<std::string, IValue>` type

jamesr66a reviewed (approved):
LGTM

jerryzh168 committed:
Update on "[quant][graphmode][refactor] getQParams return a dictionary of qparams"


Summary:
We can dictionary of quantization parameters to simplify the code
handling these things a bit

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849023](https://our.internmc.facebook.com/intern/diff/D18849023)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -21,6 +21,9 @@ namespace jit {
 namespace {
 
 using ModuleMethodVector = std::vector<std::pair<script::Module, std::string>>;
+// Map of quantization parameter name and value
```
What is used as a name here? Could you please add some details to the comment?

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -611,23 +614,21 @@ class InsertQuantDeQuantHelper {
-  std::tuple<IValue, IValue> getQParams(script::Module& module, Value* v);
+  // quantization parameters including scale, zero_point,
+  // scalar_type and axis(for per channel quantization)
+  QParamMap getQParamMap(script::Module& module, Value* v);
```
The comment describes what q-param is, not what the function does. The description of qparams should be probably moved to `using QParamMap = ...` line.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -611,23 +614,21 @@ class InsertQuantDeQuantHelper {
   void quantizeTensors(script::Module& module, Graph* g, Value* self);
 
  private:
   // TODO: we don't need to call this for each graph
```
What does this TODO mean? Is it still relevant?

jerryzh168 replied:
yeah it's still relevant. Right now we store the key for `observer_modules_to_remove_` is graph, but actually it's per-module not per-graph, so ideally we should  store a map from `ModulePtr` to observer name.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -748,23 +730,38 @@ void checkGetQParamsResult(const IValue& qparams) {
+  std::unordered_map<std::string, IValue> qparams = {
+    {"_scalar_type", scalar_type},
+  };
+  // TODO: get this from observer_module
```
What does this comment mean? Aren't we getting it from observer module when we're calling its `get_qparams` method?

jerryzh168 replied:
oh, I mean we get qscheme from `observder_module.attr("qscheme")' and from qscheme we'll know if it is per channel, I'll clarify in the comments.

ZolotukhinM reviewed (approved):
Looks good, I think some comments and the commit message might be cleaned up a bit.

jerryzh168 committed:
Update on "[quant][graphmode][refactor] getQParams return a dictionary of qparams"


Summary:
We can return a dictionary of quantization parameters in getQParams
to simplify the code that uses the quantization parameters a bit.

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849023](https://our.internmc.facebook.com/intern/diff/D18849023)

[ghstack-poisoned]
